src/modules/paymentProcessor/PP_Queue_v1.sol

```diff
@@ -0,0 +1,740 @@
+            _removeFromQueue(firstId);
+            return false;
+        }
+
+        // Check token balance and allowance.
+        if (
+            !_validTokenBalance(
+                order.order.paymentToken, order.client, order.order.amount
+            )
+        ) {
+            _updateOrderState(firstId, RedemptionState.CANCELLED);
+            return false;
+        }
+
+        return _executePaymentTransfer(firstId, order);
+    }
+
+    /// @notice	Executes the actual payment transfer for an order.
+    /// @param	orderId_ The ID of the order to process.
+    /// @param	order_ The order to process.
+    /// @return	success_ True if the payment was successful.
+    function _executePaymentTransfer(uint orderId_, QueuedOrder storage order_)
+        internal
+        returns (bool success_)
+    {
+        // Process payment.
+        (bool success, bytes memory data) = order_.order.paymentToken.call(
+            abi.encodeWithSelector(
+                IERC20(order_.order.paymentToken).transferFrom.selector,
+                order_.client,
+                order_.order.recipient,
+                order_.order.amount
+            )
+        );
+
+        // If transfer was successful.
+        if (
+            success && (data.length == 0 || abi.decode(data, (bool)))
+                && order_.order.paymentToken.code.length != 0
+        ) {
+            _updateOrderState(orderId_, RedemptionState.COMPLETED);
+            _removeFromQueue(orderId_);
+
+            // Notify client about successful payment.
+            IERC20PaymentClientBase_v1(order_.client).amountPaid(
+                order_.order.paymentToken, order_.order.amount
+            );
+
+            emit TokensReleased(
+                order_.order.recipient,
+                order_.order.paymentToken,
+                order_.order.amount
+            );
+
+            return true;
+        } else {
+            // Store as unclaimable and update state.
+            _unclaimableAmountsForRecipient[order_.client][order_
+                .order
+                .paymentToken][order_.order.recipient] += order_.order.amount;
+
+            emit UnclaimableAmountAdded(
+                order_.client,
+                order_.order.paymentToken,
+                order_.order.recipient,
+                order_.order.amount
+            );
+
+            _updateOrderState(orderId_, RedemptionState.CANCELLED);
+            _removeFromQueue(orderId_);
+
+            return false;
+        }
+    }
+
+    /// @notice	Executes all pending orders in the queue.
+    function _executePaymentQueue(address client_)
+        internal
+        clientIsValid(client_)
+    {
+        uint firstId = _queue[client_].getNextId(LinkedIdList._SENTINEL);
+        if (firstId == LinkedIdList._SENTINEL) {
+            revert Module__PP_Queue_EmptyQueue();
+        }
+
+        uint processedCount;
+        while (firstId != LinkedIdList._SENTINEL && _processNextOrder(client_))
```

Comment:
Is there no way to call `_processNextOrder` without going through this `while` loop (i.e. a from-to-approach)? I am just thinking of the case of someone spamming the queue with 1wei withdraw requests only to have this always run out of gas? Or am I missing something?

Reply:
I see the risk that you're mentioning. Do you have any suggestion how to implement this differently that it's also optimal and not too complex?

I'm not sure now what could be the workaround for this, @marvinkruse 

cc. @Zitzak